king_phisher/server/fs_utilities.py

```diff
@@ -56,6 +58,72 @@ def chown(path, user=None, group=constants.AUTOMATIC, recursive=True):
 	:type group: int, str, ``None``, :py:class:`~king_phisher.constants.AUTOMATIC`
 	:param bool recursive: Whether or not to recurse into directories.
 	"""
+
+	user, group = _resolve_information(user, group)
+
+	if recursive:
+		iterator = smoke_zephyr.utilities.FileWalker(path)
+	else:
+		iterator = (path,)
+	for path in iterator:
+		os.chown(path, user, group)
+
+
+def access(path, mode, user=None, group=constants.AUTOMATIC):
+	"""
+	This is a high-level wrapper around :py:func:`os.access` to provide
+	additional functionality. Similar to `os.access` this high-level wrapper
+	will test the given path for a variety of access modes. Additionally however,
+	*user* or *group* can be specified to test against a specific of user or group.
+
+	.. versionadded:: 1.14.0
+
+	:param str path: The path to test access for.
+	:param str mode: The mode to test access for. Set to `R_OK` to test for readability,
+		`W_OK` for writability and
+		`X_OK` to determine if path can be executed.
+	:param user: The user to test permissions for.
+	:type user: int, str, ``None``
+	:param group: The group to test permissions for. If set to
+		:py:class:`~king_phisher.constants.AUTOMATIC`, the group that *user*
+		belongs too will be used.
+	:type group: int, str, ``None``, :py:class:`~king_phisher.constants.AUTOMATIC`
+	:return: Returns ``True`` only if the user or group has the mode of permission specified else returns ``False``.
+	:rtype: bool
+	"""
+
+	user_id, group = _resolve_information(user, group)
+	file_info = os.stat(path)
+
+	if mode == 'R_OK':
+		user_permissions = stat.S_IRUSR
+		group_permissions = stat.S_IRGRP
+		other_permissions = stat.S_IROTH
+	elif mode == 'W_OK':
+		user_permissions = stat.S_IWUSR
+		group_permissions = stat.S_IWGRP
+		other_permissions = stat.S_IWOTH
+	elif mode == 'X_OK':
+		user_permissions = stat.S_IXUSR
+		group_permissions = stat.S_IXGRP
+		other_permissions = stat.S_IXOTH
+
+	if file_info.st_uid == user_id and file_info.st_mode & user_permissions:
+		return True
+
+	if file_info.st_gid == group and file_info.st_mode & group_permissions:
+		return True
+
+	for group_id in pylibc.getgrouplist(user):
+		if group_id == file_info.st_gid and file_info.st_mode & group_permissions:
+			return True
+
+	if file_info.st_mode & other_permissions:
+		return True
+
```

Comment:
Technically if `user_permissions`, `group_permissions` and `other_permissions` are all 0, then this function would return false when it should return true. This would be an edge case where the user doesn't specify anything in mode, implying a check without any permission requirements, in which case "yes the user has no or more than no permissions" so return True.